### PR TITLE
Reduce metadata configurability via CLIs

### PR DESCRIPTION
**Pull request by cgsandford**
Replaces calls to `amend_metadata` with function to update attributes only in the following CLIs (and called plugins where needed):

- improver-nowcast-extrapolate
- improver-nowcast-optical-flow
- improver-spot-extract
- improver-temp-lapse-rate

Removes metadata modification from improver-neighbour-finding CLI and associated test.

Renames `add_history_attribute` to `set_history_attribute` and removes this from where it is deeply embedded in a general utility.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/tests/nowcasting/forecasting/test_AdvectField.py [hunk 1]
```diff
@@ -240,14 +240,14 @@ def setUp(self):
         vel_y = vel_x.copy(data=2.*np.ones(shape=(4, 3)))
         vel_y.rename("advection_velocity_y")
         self.plugin = AdvectField(vel_x, vel_y)
-        self.metadata_dict = {"attributes": {
-                                 "mosg__grid_version": "1.0.0",
-                                 "mosg__model_configuration": "nc_det",
-                                 "source": "Met Office Nowcast",
-                                 "institution": "Met Office",
-                                 "title": "Nowcast on UK 2 km Standard Grid"}}
+        self.metadata_dict = {
```
Comment (bayliffe) [inline lib/improver/tests/nowcasting/forecasting/test_AdvectField.py]: attributes_dict

File: lib/improver/tests/nowcasting/forecasting/test_CreateExtrapolationForecast.py [hunk 2]
```diff
@@ -100,20 +100,19 @@ def test_basic(self):
 
     def test_basic_with_metadata_dict(self):
         """Test for simple case where __init__ does not change the input and
-           we use a metadata_dict."""
+           we amend the metadata."""
```
Comment (bayliffe) [inline lib/improver/tests/nowcasting/forecasting/test_CreateExtrapolationForecast.py]: ```amend the attributes.```

We may as well be specific.

File: lib/improver/tests/nowcasting/optical_flow/test_OpticalFlow.py [hunk 3]
```diff
@@ -645,17 +645,17 @@ def test_basic(self):
 
     def test_metadata(self):
         """Test correct output types and metadata"""
-        metadata_dict = {"attributes": {
+        metadata_dict = {
```
Comment (bayliffe) [inline lib/improver/tests/nowcasting/optical_flow/test_OpticalFlow.py]: attributes_dict

File: lib/improver/metadata/amend.py [hunk 4]
```diff
@@ -287,15 +306,6 @@ def _update_attribute(cube, attribute_name, changes, warnings_on=False):
...
             warnings.warn(msg)
-    elif "add" in changes:
-        if attribute_name in ["history"]:
-            new_history = changes
-            new_history.remove("add")
-            add_history_attribute(result, new_history[0])
-        else:
-            msg = ("Only the history attribute can be added. "
-                   "The attribute specified was {}".format(attribute_name))
-            raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/metadata/amend.py]: Yes, I'm expecting to delete that function.  I'm going to deal with what's needed for standardisation and CubeCombiner in a separate PR, where I'll get rid of whatever we're no longer using (including `_update_attributes`) and will probably wrap the whole thing in a class to manage the private methods better.  So the duplication will go.

File: lib/improver/cli/temp_lapse_rate.py [hunk 5]
```diff
@@ -177,7 +177,7 @@ def process(temperature_cube, orography_cube, land_sea_mask_cube,
                                                    land_sea_mask_cube)
     attributes = {"title": "delete", "source": "delete",
                   "history": "delete", "um_version": "delete"}
-    result = amend_metadata(result, attributes=attributes)
+    amend_attributes(result, attributes)
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/cli/temp_lapse_rate.py]: OK

**Comment by bayliffe** [comment]
improver-nowcast-optical-flow/01-help.bats is failing following the change to the help message. Please update the relevant expected output.

Otherwise this looks fine to me.

**Pull request merged**